rustup-init.ps1

```diff
@@ -0,0 +1,174 @@
+            $cputype = (Get-WmiObject Win32_OperatingSystem).OSArchitecture
+            break
+        }
+        default {
+            Write-Error "could not determine OS platform: $platform"
+        }
+    }
+
+    switch -Wildcard ($ostype) {
+        "Darwin" {
+            $ostype = "apple-darwin"
+            break
+        }
+        "Linux" {
+            $ostype = "unknown-linux-gnu"
+            break
+        }
+        "*Windows*" {
+            $ostype = "pc-windows-msvc"
+            break
+        }
+        default {
+            Write-Error "unrecognized OS type: $ostype"
+        }
+    }
+
+    switch ($cputype) {
+        # Reported by Windows 32-bit operating systems
+        "32-bit" {
+            $cputype = "i686"
+            break
+        }
+        # Reported by Windows 64-bit operating systems
+        "64-bit" {
+            $cputype = "x86_64"
+            break
+        }
+        # Reported by Darin/Linux 64-bit operating systems
+        "x86_64" {
+            break
+        }
+        default {
+            Write-Error "unknown CPU type: $cputype"
+        }
+    }
+
+    $arch = "${cputype}-${ostype}"
+    Write-Debug "Detected arch '$arch'"
+    return "$arch"
+}
+
+# If RUSTUP_UPDATE_ROOT is unset, default it.
+function Get-RustupUpdateRoot {
+    if (Test-Path env:\RUSTUP_UPDATE_ROOT) {
+        return "$env:RUSTUP_UPDATE_ROOT"
+    } else {
+        return "https://static.rust-lang.org/rustup"
+    }
+}
+
+function Install-Rustup([AllowEmptyCollection()] [string[]]$Argv) {
+    $arch = Get-Architecture
+    switch -Wildcard ($arch) {
+        "*windows*" { $ext = ".exe"; break }
+        default { $ext = "" }
+    }
+
+    $url = "$(Get-RustupUpdateRoot)/dist/$arch/rustup-init$ext"
+
+    $dir = New-TemporaryDirectory
+    $file = Join-Path "$dir" "rustup-init$ext"
+
+    try {
+        if ("$arch" -match "windows") { Set-HighestEncryption }
+
+        Write-Debug "Downloading $url to $file"
+        (New-Object System.Net.WebClient).DownloadFile($url, $file)
+        if (-Not ("$arch" -match "windows")) { chmod u+x "$file" }
+
+        Write-Debug "Calling '$file $Argv'"
+        & "$file" @Argv
+    } finally {
+        Write-Debug "Cleaning $dir"
+        Remove-Item "$dir" -Force -Recurse
+    }
+}
+
+function New-TemporaryDirectory {
+    $parent = [System.IO.Path]::GetTempPath()
+    [string]$name = [System.Guid]::NewGuid()
+    New-Item -ItemType Directory -Path (Join-Path "$parent" "$name")
+}
+
+function Set-HighestEncryption {
+    # Implementation adapted from the Chocolatey installer script,
+    # see: https://chocolatey.org/install.ps1
+
+    # Attempt to set highest encryption available for SecurityProtocol.
+    # PowerShell will not set this by default (until maybe .NET 4.6.x). This
+    # will typically produce a message for PowerShell v2 (just an info message
+    # though)
+    try {
+        # Set TLS 1.2 (3072), then TLS 1.1 (768), then TLS 1.0 (192), finally
```

Comment:
Please don't set TLS 1.1 or TLS 1.0 and definitely *NEVER* SSL 3.0 fallbacks.

If the minimum .NET supported version is sufficiently high that this will be awkward then instead try and set `1.2` and failing that, report the warning and continue as you do below.  .NET 4.5 is nearly seven years old, so if that's all that's needed, I'd be tempted to simply mandate it and error rather than warn below.

Reply:
Now *that* is something I can dial back. Admittedly, this was heavily inspired by Chocolatey's `install.ps1` and I suspect that they are targeting a much wider set of install targets. Considering anyone installing Rust may likely already have Chocolatey installed (and therefore could install more software), this might be overkill. In fact, it might make sense to remove that function completely and wait for any issues that would prompt taking action. Make sense?

Reply:
I think you should attempt to explicitly set TLS 1.2 and fail if that can't be done.

Reply:
Also...You should be a good citizen here.  Changing that method will affect the user AFTER you run this script.

Instead you should 1) set it, 2) do the download and then 3) restore the previous value.

Reply:
Right--I was burning myself earlier when I had `exit` calls in the script. Did not work so well when you pulled that into a scriptblock for execution--I couldn't understand why my session…exited 😄 

I'll see what I can do here…

Reply:
Trying to find some prior art for you ....

Reply:
I couldn't find it but basically

``` powershell
$OldSPM = [System.Net.ServicePointManager]::SecurityProtocol
try {
   <do lots of stuff here>
  [System.Net.ServicePointManager]::SecurityProtocol = <new value>

  <Download what you need>
} finally {
  [System.Net.ServicePointManager]::SecurityProtocol = $OldSPM
}
```

Reply:
Okay, I went for TLS 1.2 and a fail if that didn't succeed. I also used the save-try-finally-restore example above and feel much better about that.